File: src/main/java/shopping/application/member/MemberService.java
package shopping.application.member;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import shopping.domain.member.Member;
import shopping.domain.member.exception.LoginFailedException;
import shopping.domain.repository.MemberRepository;
import shopping.dto.LoginRequest;
import shopping.dto.LoginResponse;
import shopping.dto.SignUpRequest;
import shopping.infrastructure.auth.JwtTokenProvider;

@Service
@Transactional
public class MemberService {
    private final MemberRepository memberRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;

    public MemberService(MemberRepository memberRepository, PasswordEncoder passwordEncoder, JwtTokenProvider jwtTokenProvider) {
        this.memberRepository = memberRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
    }

    public Long signUp(SignUpRequest request) {
        if(memberRepository.existByEmail(request.email())) {
            throw new IllegalArgumentException(request.email());
        }

        String encodePassword = passwordEncoder.encode(request.password());
        Member member = Member.create(request.email(), encodePassword);

        return memberRepository.save(member).getId();
    }

    @Transactional(readOnly=true)
    public LoginResponse login(LoginRequest request) {
        Member member = memberRepository.findByEmail(request.email())
                .orElseThrow(() -> new LoginFailedException(request.email()));

        if(!passwordEncoder.matches(request.password(), member.getPassword())) {
            throw new LoginFailedException(request.email());
        }

        String token = jwtTokenProvider.createToken(member.getEmail(), member.getRole());

        return new LoginResponse(token);
    }
}

File: src/main/kotlin/shopping/application/member/MemberService.kt
package shopping.application.member

import shopping.domain.member.Member
import shopping.domain.member.MemberRepository
import shopping.support.error.CoreException
import org.springframework.stereotype.Service
import shopping.support.error.ErrorType

@Service
class MemberService(
    private val memberRepository: MemberRepository
) {
    fun register(email: String, password: String): Member {
        if (memberRepository.existsByEmail(email)) {
            throw CoreException(ErrorType.CONFLICT, "사용자가 이미 존재합니다.")
        }
        return memberRepository.save(
            Member(
                email = email,
                password = password
            )
        )
    }

    fun login(email: String, password: String) : Long {
        val member = memberRepository.findByEmail(email)
        ?: throw CoreException(ErrorType.BAD_REQUEST, "사용자 존재하지 않습니다.")

        if (!member.checkPassword(password)) {
            throw CoreException(ErrorType.BAD_REQUEST, "비밀번호가 틀렸습니다.")
        }
        return member.id;
    }
}